dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,517 @@
+    Generator,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Type,
+    TypeVar,
+    Union,
+)
+
+
+def flatten(maybe_iterables: Any) -> Iterator:
+    """
+    flattens `maybe_iterables` by descending into items that are Iterable
+    """
+
+    def single_item_generator(item: Any) -> Generator[Any, None, None]:
+        yield item
+
+    if not isinstance(maybe_iterables, (list, tuple, set, Iterator)):
+        return single_item_generator(maybe_iterables)
+    return chain.from_iterable(
+        (flatten(maybe_iterable) for maybe_iterable in maybe_iterables)
+    )
+
+
+class Node(ABC):
+    """
+    Base class for all DJ AST nodes.
+
+    DJ nodes are python dataclasses with the following patterns:
+        - Attributes are either
+            - primitives (int, float, str, bool, None)
+            - iterable from (list, tuple, set)
+            - Enum
+            - descendant of `Node`
+        - Attributes starting with '_' are "obfuscated" and are not included in `children`
+
+    """
+
+    _parents: Optional[Set["Node"]]
+
+    @property
+    def parents(self) -> Set["Node"]:
+        """
+        get the parents of the node
+        """
+        try:
+            return self._parents  # type: ignore
+        except AttributeError:
+            self._parents = set()
+            return self._parents
+
+    def add_parents(self, *parents: "Node") -> "Node":
+        """
+        add parents to the node
+        """
+        for parent in parents:
+            self.parents.add(parent)
+        return self
+
+    def add_self_as_parent(self) -> "Node":
+        """
+        adds self as a parent to all children
+        """
+        for child in self.children:
+            child.add_parents(self)
+        return self
+
+    def flatten(self) -> Iterator["Node"]:
+        """
+        flatten the sub-ast of the node as an iterator
+        """
+        return self.filter(lambda _: True)
+
+    def fields(
+        self,
+        flat: bool = True,
+        nodes_only: bool = True,
+        obfuscated: bool = False,
+        nones: bool = False,
+    ) -> Iterator:
+        """
+        Args:
+            flat: return a flattened iterator (if children are iterable)
+            nodes_only: do not yield children that are not Nodes (trumped by `obfuscated`)
+            obfuscated: yield fields that have leading underscores
+                (typically accessed via a property)
+            nones: yield values that are None
+                (optional fields without a value); trumped by `nodes_only`
+
+        Returns:
+            Iterator: returns all children of a node given filters
+                and optional flattening (by default Iterator[Node])
+        """
+        child_generator = iter(
+            self.__dict__[field.name]
+            for field in fields(self)
+            if (not field.name.startswith("_") if not obfuscated else True)
+        )  # exclude obfuscated fields if `obfuscated` is True
```

Comment:
I'd split this to increase readability a little bit, and you don't need the `iter()`:

```suggestion
        child_generator = (self.__dict__[field.name] for field in fields(self))
        if not obfuscated:
            child_generator = (field for field in child_generator if not field.name.startswith("_"))
```